pycbc/coordinates/space.py

```diff
@@ -0,0 +1,938 @@
+
+    return params_ssb
+
+
+def rotation_matrix_ssb_to_geo(epsilon=np.deg2rad(23.439281)):
+    """ The rotation matrix (of frame basis) from SSB frame to
+    geocentric frame.
+
+    Parameters
+    ----------
+    epsilon : float
+        The Earth's axial tilt (obliquity), in the unit of 'radian'.
+
+    Returns
+    -------
+    r : numpy.array
+        A 3x3 rotation matrix from SSB frame to geocentric frame.
+    """
+    r = Rotation.from_rotvec([
+        [-epsilon, 0, 0]
+    ]).as_matrix()
+
+    return np.array(r[0])
+
+
+def earth_position_ssb(t_geo):
+    """ Calculating the position vector and angular displacement of the Earth
+    in the SSB frame, at a given time. By using Astropy.
+
+    Parameters
+    ----------
+    t_geo : float
+        The time when a GW signal arrives at the origin of geocentric frame,
+        or any other time you want.
+
+    Returns
+    -------
+    (p, alpha) : tuple
+    p : numpy.array
+        The position vector of the Earth in the SSB frame. In the unit of 'm'.
+    alpha : float
+        The angular displacement of the Earth in the SSB frame.
+        In the unit of 'radian'.
+    """
+    t = Time(t_geo, format='gps')
+    pos = get_body_barycentric('earth', t)
+    # BarycentricMeanEcliptic doesn't have obstime attribute,
+    # it's a good inertial frame, but ICRS is not.
+    icrs_coord = SkyCoord(pos, frame='icrs', obstime=t)
+    bme_coord = icrs_coord.transform_to(
+                    BarycentricMeanEcliptic(equinox='J2000'))
+    x = bme_coord.cartesian.x.to(units.m).value
+    y = bme_coord.cartesian.y.to(units.m).value
+    z = bme_coord.cartesian.z.to(units.m).value
+    p = np.array([[x], [y], [z]])
+    alpha = bme_coord.lon.rad
+
+    return (p, alpha)
+
+
+def t_geo_from_ssb(t_ssb, longitude_ssb, latitude_ssb, use_astropy, frame):
+    """ Calculating the time when a GW signal arrives at the barycenter
+    of the Earth, by using the time and sky localization in SSB frame.
+
+    Parameters
+    ----------
+    t_ssb : float
+        The time when a GW signal arrives at the origin of SSB frame.
+        In the unit of 's'.
+    longitude_ssb : float
+        The ecliptic longitude of a GW signal in SSB frame.
+        In the unit of 'radian'.
+    latitude_ssb : float
+        The ecliptic latitude of a GW signal in SSB frame.
+        In the unit of 'radian'.
+
+    Returns
+    -------
+    t_geo : float
+        The time when a GW signal arrives at the origin of geocentric frame.
+    """
+    k = localization_to_propagation_vector(
+            longitude_ssb, latitude_ssb, use_astropy, frame)
+
+    def equation(t_geo):
+        # Earth is moving, when GW arrives at Earth center,
+        # time is t_geo, not t_ssb.
+        p = earth_position_ssb(t_geo)[0]
+        return t_geo - t_ssb - np.vdot(k, p) / c.value
+
+    return fsolve(equation, t_ssb)[0]
+
+
+def t_ssb_from_t_geo(t_geo, longitude_ssb, latitude_ssb,
+                     use_astropy=True, frame=None):
```

Comment:
Is the 'use_astropy' option now working either way? If so, maybe this should be part of the unittest to make sure given it gave so many troubles before? Or should we settle on always using astropy?

Reply:
@ahnitz In the unittest, using astropy is the default.